compiler/tests/evm_runtime.rs

```diff
@@ -0,0 +1,170 @@
+#![cfg(feature = "solc-backend")]
+use fe_compiler::yul::runtime::functions;
+use yultsur::*;
+
+mod utils;
+use utils::*;
+
+macro_rules! assert_eq {
+    ($a:tt, $b:tt) => {
+        statement! {
+            (if (iszero((eq($a, $b)))) {
+                (revert(0, 0))
+            })
+        }
+    };
+}
+
+#[test]
+fn test_alloc_and_avail() {
+    with_executor(&|mut executor| {
+        test_runtime_functions(
+            &mut executor,
+            functions::std(),
+            statements! {
+                (let a := avail())
+                (let b := alloc(5))
+                (let c := alloc(10))
+                (let d := avail())
+
+                [assert_eq!(b, a)]
+                [assert_eq!(c, (add(b, 5)))]
+                [assert_eq!(d, (add(c, 10)))]
+            },
+        );
+    })
+}
+
+#[test]
+fn test_mcopys() {
+    with_executor(&|mut executor| {
+        test_runtime_functions(
+            &mut executor,
+            functions::std(),
+            statements! {
+                (let a := 26)
+                (let b := 27)
+
+                (mstore(42, a))
+                (mstore(74, b))
+                (mcopys(42, 100, 64))
+
+                [assert_eq!(a, (sload(100)))]
+                [assert_eq!(b, (sload(132)))]
+            },
+        );
+    })
+}
+
+#[test]
+fn test_scopym() {
+    with_executor(&|mut executor| {
+        test_runtime_functions(
+            &mut executor,
+            functions::std(),
+            statements! {
+                (let a := 26)
+                (let b := 27)
+
+                (sstore(42, a))
+                (sstore(74, b))
+
+                (let ptr1 := scopym(42, 64))
+                (let ptr2 := add(ptr1, 32))
+
+                [assert_eq!(a, (mload(ptr1)))]
+                [assert_eq!(b, (mload(ptr2)))]
+            },
+        );
+    })
+}
+
+#[test]
+fn test_mloadn() {
+    with_executor(&|mut executor| {
+        test_runtime_functions(
+            &mut executor,
+            functions::std(),
+            statements! {
+                (let a := 0x4200000000000000000000000000000000000000000000000000000000420026)
+                (let b := 0x0000000000000000000000000000000000000000000000000000000000000026)
+                (mstore(100, a))
+
+                [assert_eq!(b, (mloadn(130, 2)))]
+            },
+        );
+    })
+}
+
+#[test]
+fn test_sloadn() {
+    with_executor(&|mut executor| {
+        test_runtime_functions(
+            &mut executor,
+            functions::std(),
+            statements! {
+                (let a := 0x4200000000000000000000000000000000000000000000000000000000420026)
+                (let b := 0x0000000000000000000000000000000000000000000000000000000000000026)
+                (sstore(100, a))
+
+                // this fails for some reason.. possible bug?
+                // [assert_eq!(b, (sloadn(130, 2)))]
```

Comment:
it looks like this might be a bug @cburgdorf. I could be missing something tho

Reply:
This is interesting. You can prove `[assert_eq!(0x42, (sloadn(100, 1)))]` but that's as far as I got. I tried changing the second byte to anything but `00` and then use `sloadn(101, 1)` but that already fails. So, maybe a base assumption about how storage works is wrong? :sweat_smile:

Reply:
I'm going to merge this and continue digging into the storage stuff. Pretty cool that we already found an issue with this testing.